### ISSUE TITLE
BrowserAnimationsModule is not tree-shakable

### ISSUE THREAD

**Issue by ranma42** [angular/angular#36306]
# 🐞 bug report

### Affected Package

The issue happens when using the `@angular/platform-browser/animations` module, but it is not obvious whether the actual problem is in there or in the compiler.

### Is this a regression?

No.

### Description

An unused module should not affect the results of the build; instead the size of main.js grows when importing `BrowserAnimationsModule` in an application that does not use it.

## 🔬 Minimal Reproduction

The issue cannot be shown using StackBlitz, but the repository https://github.com/ranma42/treeshake-material can be used to reproduce it.

The steps to reproduce the issue are illustrated in the commit messages of the repository. Namely:

 1. Create a new application with ng new treeshake-material with default settings (no routing, CSS style).
 2. Install @angular/material with default settings (indigo-pink theme, no typography styles, yes to browser animations); it will automatically import `BrowserAnimationsModule` in `AppModule`.

As show in the commit messages of the repository, `main.js` grows from 137 KB to 207 KB.

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 9.1.0
Node: 12.16.0
OS: darwin x64

Angular: 9.1.0
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.901.0
@angular-devkit/build-angular     0.901.0
@angular-devkit/build-optimizer   0.901.0
@angular-devkit/build-webpack     0.901.0
@angular-devkit/core              9.1.0
@angular-devkit/schematics        9.1.0
@angular/cdk                      9.2.0
@angular/material                 9.2.0
@ngtools/webpack                  9.1.0
@schematics/angular               9.1.0
@schematics/update                0.901.0
rxjs                              6.5.4
typescript                        3.8.3
webpack                           4.42.0
</code></pre>

**Anything else relevant?**
According to the [Ivy compatibility guide](https://angular.io/guide/ivy-compatibility#payload-size-debugging), modules should be tree-shaken when marked side-effect-free.
`@angular/platform-browser/animations` is declared [side-effect-free](https://github.com/angular/angular/blob/master/packages/platform-browser/animations/package.json#L11), but it does not seem to suffice.

See also https://github.com/angular/components/issues/18944 for a similar issue affecting `@angular/material`

**Comment by JoostK** [comment]
There is a key part in this sentence in the Ivy compatibility guide that is essential to the tree-shakeability of an ECMAScript module:

> This will ensure that the libraries will be properly tree-shaken if they are imported **but not directly referenced**.

In your case, the NgModules you import are added to `@NgModule.imports`, which means there's a direct reference to the imported NgModule's class so it is retained in the final bundle.

The story is unfortunately a bit complicated in Angular's case. First, there's two kinds of modules:

1. ECMAScript modules, i.e. the Javascript files that use `import`/`export` statements like `import {Foo} from './foo';`.
2. Angular's NgModules, which also has imports and exports of other NgModules.

The `sideEffects` flag that the guide mentions affects the available optimizations that an optimizer tool can apply. When the flag is set to `false`, the library indicates that importing it will not execute any logic that may influence the rest of the environment, e.g. set some global variables, so the optimizer can be more aggressive in its optimizations. However, any classes/functions/symbols that are referenced from the application will not be dropped by a tree-shaker, as it would invalidate the reference.

Therefore, referencing an imported NgModule class such as `BrowserAnimationsModule` will prevent it from being tree-shaken. Having the `BrowserAnimationsModule` as an NgModule import will in fact influence how your application behaves, as an imported NgModule can provide services using the `@NgModule.providers` field. As such, it would be invalid for a tree-shaker to remove the class wholesale, as you'd end up without the providers.

---

Unfortunately, it does get more complicated than that. Any _declarations_ that you reference in `@NgModule.declarations` **are** eligible for tree shaking, due to a particular design choice of Ivy's compiled code. Although you create a reference to a component/directive/pipe class in `@NgModule.declarations`, none of these references are retained into the compiled code. The information is _only_ used during compilation of the project to tell what other declarations can be used within a component's template. The declarations that are used in a given template will be referenced from the compiled template code, but no longer from the NgModule. This design is chosen as it does allow for tree-shaking any declaration that is never used from within a template. This is only valid because referencing a class in `NgModule.declarations` does not have any side-effects other than defining a component's scope, but this information is redundant when all components are AOT compiled.

---

In conclusion, there is no bug here with `BrowserAnimationsModule` nor with the modules reported in https://github.com/angular/components/issues/18944.

I am wondering if you feel that the documentation would need clarification here. So I'll keep this open as a documentation request. Please follow up with your expectations here.

**Comment by ranma42** [comment]
I eventually found https://github.com/angular/angular-cli/pull/16794 which enabled me to inspect the build output and confirm that:

 1. components and directives are being removed by minification (as desired)
 2. modules are not being removed (IIUC this is expected behaviour, as they are directly referenced in the app code or in some other referenced modules)

I would also like to point out that there are also a bunch of services are not tree-shaken even though they are apparently tree-shakeable.
For example I believe [FocusMonitor](https://github.com/angular/components/blob/204db567c0a4352cfe08a6eea63ea57f58a8d8e3/src/cdk/a11y/focus-monitor/focus-monitor.ts#L59-L60) is retained because it is also listed as a [provider](https://github.com/angular/components/blob/204db567c0a4352cfe08a6eea63ea57f58a8d8e3/src/cdk/tree/tree-module.ts#L31)).
I guess the same happens for [`IterableDiffers`](https://github.com/angular/angular/blob/86104b82b8af87efab46db50962e6b16f3724672/packages/core/src/change_detection/differs/iterable_differs.ts#L144-L150) provided by [`ApplicationModule`](https://github.com/angular/angular/blob/58f10026c4d1d55a96f517ee721ac694602a34c4/packages/core/src/application_module.ts#L92), but I am much less confident since the syntax used in `IterableDiffers` is apparently required to let it work even if no injector is available.

On the docs side, "Payload size debugging" would be much easier if the guide provided some information about how to inspect the build output, for example by documenting `NG_BUILD_DEBUG_OPTIMIZE`.
If this is not feasible (in the PR this is mentioned as "internal api" so it might be inappropriate for long-term commitment), it might be a good idea to add instrumentation to the compiler to conditionally emit the output of the intermediate translation/optimization steps. This kind of flags and hooks is very valuable when trying to figure out the details of the compilation and is, in fact, available on many compilation tools.

> Therefore, referencing an imported NgModule class such as BrowserAnimationsModule will prevent it from being tree-shaken. Having the BrowserAnimationsModule as an NgModule import will in fact influence how your application behaves, as an imported NgModule can provide services using the @NgModule.providers field. As such, it would be invalid for a tree-shaker to remove the class wholesale, as you'd end up without the providers.

IIUC, adding a module to `@NgModule.declarations` (without any other explicit code change) can affect the behaviour of the application by:
 - overriding an existing provider
 - performing operations on "external" providers upon module instantiation

It would be convenient to have some feedback about such modules and providers. Is there any way to dump the DI tree?
It is not clear to me whether "somewhat-treeshakeable" services like the `FocusMonitor` above are expected and what would change if they were dropped from `@NgModule.providers` (leaving `providedIn: 'root'`). If this is a safe change, maybe the compiler could warn about it.

It would be awesome if modules were emitted in a tree-shakeable form, i.e. the same approach applied to `@NgModule.declarations` was also applied to `@NgModule.providers`.
This would enable minification of code such as:
```js
const CDK_CONNECTED_OVERLAY_SCROLL_STRATEGY_PROVIDER = {
    provide: new InjectionToken("cdk-connected-overlay-scroll-strategy"),
    deps: [ overlay_Overlay ],
    useFactory: function(overlay) {
        return () => overlay.scrollStrategies.reposition();
    }
};
```
that is currently retained in the build output, even though it is obviously dead code (the minifier inlined the token definition, since it was not used elsewhere).

As a side-effect, this would possibly enable the minification of no-op modules, that are currently retained even if they contain nothing at all. For example [`BidiModule`](https://github.com/angular/components/blob/8050f633b56b6c048fc72dad2ab79304afdfad19/src/cdk/bidi/bidi-module.ts) is built into an empty module:
```js
bidi_BidiModule = (() => {
    class BidiModule {}
    return BidiModule.\u0275mod = \u0275\u0275defineNgModule({
        type: BidiModule
    }), BidiModule.\u0275inj = \u0275\u0275defineInjector({
        factory: function(t) {
            return new (t || BidiModule);
        }
    }), BidiModule;
})()
```

I believe this issue can be closed as it is "working as expected", but I would like to open issues (feature requests) and (when possible) PRs for the points I mentioned.
I would love some guidance into which ones are actionable:
 1. add documentation about `NG_BUILD_DEBUG_OPTIMIZE`
 2. add flags to emit intermediate compilation output
 3. add flags to warn about "inconsistently-provided" services
 3. add flags (or command) to emit the DI tree of a module
 4. translate `NgModule`s to a tree-shakable form

**Comment by hoang-innomizetech** [comment]
@JoostK I am not sure whether my issue relate to this but I am trying to optimize the webpack by adding `sideEffect` flag to the package.json. After building AOT bundler, my app cannot be loaded, it just display a blank page without any error. Do you know what can be the root cause and any resolution to fix that issue? 

Step to replicate my issue
* Generate latest angular app
* Add _sideEffect: false_ to the _package.json_
* _ng build --prod_
* Serve with _http-server_ and the app display a blank page

FYI, If I disabled optimization by adding the environment variable that @ranma42 mentioned above, then the app is loadable  i..e _NG_BUILD_DEBUG_OPTIMIZE=* ng build --prod_

**Comment by kapunahelewong** [comment]
Status: Docs team has enough material above to make the docs update.

**Comment by sod** [comment]
@hoang-innomizetech `sideEffects: false` is tricky. E.g. imports like `import 'foo/bar'` are being removed in sideEffects free mode. This may prevent zone.js from being included in the build.

see https://webpack.js.org/guides/tree-shaking/

**Comment by JeanMeche** [comment]
Somehow related, this will allow to lazy loaded the animations : #50738

**Comment by alxhub** [comment]
Importing an NgModule _is_ directly using it - adding `BrowserAnimationsModule` to your application enables animations in any components that happen to be using them. The new ability to lazy load it though is really nice!

**Comment by ranma42** [comment]
@alxhub would it make sense for any of the points I suggested (such as documenting ways to debug/analyze the DI and tree-shaking) to be opened as feature requests?
For example I cannot find `NG_BUILD_DEBUG_OPTIMIZE` documented anywhere; I only find hits in github issues plus a mention [here](https://github.com/angular/angular/tree/5978b3d132b405b8df76216c63f0520b485c2b46/integration), which looks like "internal documentation"

**Comment by JeanMeche** [comment]
@ranma42 This is a request for the CLI repo which is handling those flags. 
Also I would more recommend `NG_BUILD_MANGLE=0` which is more command I achieves what you are looking for.
